packages/teleport/src/Discover/LoginTrait/useLoginTrait.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2022 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect } from 'react';
+import useAttempt from 'shared/hooks/useAttemptNext';
+import type { User } from 'teleport/services/user';
+import { DiscoverContext } from '../discoverContext';
+import type { AgentStepProps } from '../Shared';
+
+export function useLoginTrait({ ctx, props }: Props) {
+  const [user, setUser] = useState<User>();
+  const { attempt, run, setAttempt, handleError } = useAttempt('processing');
+  const [loginMap, setLoginMap] = useState<Record<string, boolean>>({});
+
+  useEffect(() => {
+    run(() =>
+      ctx.userService.fetchUser(ctx.username).then(user => {
+        setUser(user);
+        let loginMap = {};
+        user.traits.logins.forEach(login => (loginMap[login] = true));
+        setLoginMap(loginMap);
+      })
+    );
+  }, []);
+
+  function nextStep(logins: string[]) {
+    setAttempt({ status: 'processing' });
+    ctx.userService
+      .updateUser({
+        ...user,
+        traits: { ...user.traits, logins },
+      })
+      .then(props.nextStep)
+      .catch(handleError);
+  }
```

Comment:
This could use async/await for a nicer flow

```typescript
  async function nextStep(logins: string[]) {
    setAttempt({ status: 'processing' });
    
    try {
      const user = await ctx.userService.updateUser({ ...user, etc });
      
      props.nextStep(user);
    } catch (err) {
      handleError(err);
    }
  }
```

Reply:
tbh, i think the chaining is cleaner and easier to read :sweat_smile:  (maybe it's b/c i'm so used to it). but i am open to discussion with the team to vote for writing things a certain way for consistency sake, otherwise i'm keeping it as is atm

Reply:
Ah see I find async/await easier to read, and it's more flexible when doing more with promises - https://dev.to/gafi/7-reasons-to-always-use-async-await-over-plain-promises-tutorial-4ej9

Happy for the team to chime in on this

Reply:
Let's leave this for a discussion out of the scope of this PR as convention in the project is to use chaining.